### PR TITLE
Allow the mig resource controllers to watch a specific namespace

### PR DESCRIPTION
**Pull request by mansam**
Following discussion with @jortel and @dymurray this splits the CAM role into individual roles for each of the mig resource controllers, and updates the watch predicates to filter based on the SANDBOX_NAMESPACE environment variable if it's present. The cluster and plan controllers also use a SINGLETON_NAMESPACE environment variable, if present, to restrict where it will look for cluster updates.

### CODE REVIEW

**Comment by dymurray** [comment]
@mansam FYI please target this for the non-admin feature branch. I am about to rebase it with master

**Comment by djwhatle** [comment]
Switching non-admin PRs to use non-admin label for tracking.

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -31,19 +31,12 @@ import (
...
-	migmigration.Add,
-	migstorage.Add,
-	migplan.Add,
-}
-
-//
-// List of controller add functions for the Discovery role.
-var DiscoveryControllers = []AddFunction{
-	discovery.Add,
+var Controllers = map[string]AddFunction{
```
Comment (dymurray) [inline pkg/controller/controller.go:34]: :+1:

File: pkg/controller/migcluster/predicate.go [hunk 2]
```diff
@@ -14,6 +15,9 @@ type ClusterPredicate struct {
 }
 
 func (r ClusterPredicate) Create(e event.CreateEvent) bool {
+	if !common.IsInSingletonNamespace(e.Meta.GetNamespace()) {
```
Comment (dymurray) [inline pkg/controller/migcluster/predicate.go:18]: Can you help explain to me what exactly this is doing? I'm not super familiar with the predicate code.

File: pkg/controller/common/common.go [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package common
```
Comment (jortel) [inline pkg/controller/common/common.go:1]: Let's avoid package, .go files and functions named: "util", "common", "helper".  They lead to _buckets of functions_. I strongly believe that any functionality added should be scoped and grouped by concern.

These functions look like a good fit in the _settings_ package and likely a method on `_Settings`.

File: pkg/controller/common/common.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+import "github.com/konveyor/mig-controller/pkg/settings"
+
+func IsInSingletonNamespace(namespace string) bool {
```
Comment (jortel) [inline pkg/controller/common/common.go:5]: Perhaps we can find a more descriptive name than "Singleton".  Perhaps "Privileged"?

File: pkg/controller/controller.go (hunk 1 above)
Comment (jortel) [reply pkg/controller/controller.go:34]: :+1:

File: pkg/controller/common/common.go (hunk 3 above)
Comment (mansam) [reply pkg/controller/common/common.go:1]: Sounds good.

File: pkg/controller/common/common.go (hunk 4 above)
Comment (mansam) [reply pkg/controller/common/common.go:5]: Privileged sounds much better to me. @dymurray?

Comment (sseago) [reply pkg/controller/common/common.go:5]: "Singleton" generally refers to an object for which only one is able to exist its type/class. There isn't generally a notion of "privileged" there. If privileges are relevent, then I'd say "PrivilegedNamespace" makes more sense.

**Comment by jortel** [comment]
It seem reasonable to finalize the _admin_ vs _non-admin_ installation as either: In the sandbox or in the privileged namespaces.  As such, I think the code simply needs to answer the question: "_Are we in the sandbox or not."_

Suggest something like:

In settings:
- SandboxNamespace
- PrivilegedNamespace

```
+ func (s *_Settings) InSandbox(r MigResource) bool {
    return r.GetNamespace() == s.SandboxNamespace
}
```

Using the interface:
```
type MigResource interface {
+ InSandbox() bool
}
...
func (r *MigPlan) InSandbox() bool {
    return settings.Settings.InSandbox(r)
```
In the predicates:

```
if cluster.InSandbox() {
    return false
}
...
if !plan.InSandbox() {
    return false
}
```

File: pkg/controller/common/common.go (hunk 4 above)
Comment (dymurray) [reply pkg/controller/common/common.go:5]: @mansam yes `privileged` sounds good to me. Or `admin`.

**Comment by mansam** [comment]
\test

**Comment by mansam** [comment]
\test-with-operator 312

**Comment by mansam** [comment]
\test-with-operator 312

**Comment by mansam** [comment]
\test-with-operator 312

**Comment by mansam** [comment]
\test-with-operator 312

**Comment by mansam** [comment]
\test-with-operator 312

**Comment by mansam** [comment]
\test-with-operator 312

**Comment by mansam** [comment]
\test-with-operator 312

**Comment by dymurray** [comment]
Fixes #471

**Pull request merged**